semgrep-core/src/core/ast/Find_target.ml

```diff
@@ -0,0 +1,68 @@
+(*
+   Find and filter targets.
+*)
+
+open Common
+module Resp = Semgrep_core_response_t
+
+let exclude_files_in_skip_lists roots =
+  let paths, skipped_paths =
+    Skip_code.filter_files_if_skip_list ~root:roots roots
+  in
+  let skipped =
+    skipped_paths
+    |> List.map (fun path ->
+           {
+             Resp.path;
+             reason = Excluded_by_config;
+             details = "excluded by 'skip list' file";
+             skipped_rule = None;
+           })
+  in
+  (paths, skipped)
+
+(*
+   Some source files are really huge (> 20 MB) and they cause
+   some annoying 'out of memory' crash that sometimes even the use
+   of mem_limit_mb above does not solve.
+
+   We could configure the size limit based on a per-language basis if we
+   know that a language parser can handle larger files.
+*)
+let filter_by_size _lang paths =
+  let max_bytes = 5_000_000 in
+  paths
+  |> Common.partition_result (fun path ->
+         let size = Common2.filesize path in
+         if size > max_bytes then
+           Error
+             {
+               Resp.path;
+               reason = Too_big;
+               details =
+                 spf "target file size exceeds %i bytes at %i bytes" max_bytes
+                   size;
+               skipped_rule = None;
+             }
+         else Ok path)
+
+let files_of_dirs_or_files ?(keep_root_files = true) lang roots =
+  let explicit_targets, paths =
+    if keep_root_files then
+      roots
+      |> List.partition (fun path ->
+             Sys.file_exists path && not (Sys.is_directory path))
+    else (roots, [])
+  in
+  let paths = Common.files_of_dir_or_files_no_vcs_nofilter paths in
+  let paths, skipped1 = exclude_files_in_skip_lists paths in
+  let paths, skipped2 = Guess_lang.inspect_files lang paths in
+  let paths, skipped3 = filter_by_size lang paths in
+  let skipped = List.flatten [ skipped1; skipped2; skipped3 ] in
```

Comment:
`List.flatten` is not tail-rec, if we scan a very large repos with thousands of files that need to be skipped, this may stack overflow perhaps?

Reply:
We have `Common.flatten` which is tail-rec. Here we could even use some `rev_flatten` since you're going to sort `skipped` anyways.

Reply:
yes. Btw, we should automate this with semgrep.